File: api/internal/core/dto/connection.go
package dto

import "github.com/huynhanx03/datadock/internal/core/entity"

type ConnectionInput struct {
	WorkspaceID     string         `json:"workspaceId" binding:"required"`
	Name            string         `json:"name" binding:"required,max=100"`
	Engine          entity.Engine  `json:"engine" binding:"required"`
	Host            string         `json:"host" binding:"required,max=255"`
	Port            int            `json:"port" binding:"gte=0,lte=65535"`
	Database        string         `json:"database" binding:"max=255"`
	Username        string         `json:"username" binding:"max=255"`
	Password        string         `json:"password"`
	ClearPassword   bool           `json:"clearPassword"`
	SSLMode         entity.SSLMode `json:"sslMode"`
	SSLCAPath       string         `json:"sslCaPath" binding:"max=4096"`
	SSLCertPath     string         `json:"sslCertPath" binding:"max=4096"`
	SSLKeyPath      string         `json:"sslKeyPath" binding:"max=4096"`
	ProxyURL        string         `json:"proxyUrl" binding:"max=2048"`
	ClearProxyAuth  bool           `json:"clearProxyCredentials"`
	SSHTunnel       SSHTunnelInput `json:"sshTunnel"`
	ReadOnly        bool           `json:"readOnly"`
	AutoReconnect   bool           `json:"autoReconnect"`
	MaxOpenConns    int            `json:"maxOpenConns" binding:"gte=0,lte=100"`
	MaxIdleConns    int            `json:"maxIdleConns" binding:"gte=0,lte=100"`
	ConnMaxLifetime int            `json:"connMaxLifetimeSeconds" binding:"gte=0,lte=86400"`
	ConnMaxIdleTime int            `json:"connMaxIdleTimeSeconds" binding:"gte=0,lte=86400"`
}

type SSHTunnelInput struct {
	Enabled        bool   `json:"enabled"`
	Host           string `json:"host" binding:"max=255"`
	Port           int    `json:"port" binding:"gte=0,lte=65535"`
	Username       string `json:"username" binding:"max=255"`
	Password       string `json:"password"`
	ClearPassword  bool   `json:"clearPassword"`
	PrivateKeyPath string `json:"privateKeyPath" binding:"max=4096"`
	KnownHostsPath string `json:"knownHostsPath" binding:"max=4096"`
}

type ConnectionView struct {
	entity.ConnectionRuntimeStatus
	ID              string         `json:"id"`
	WorkspaceID     string         `json:"workspaceId"`
	Name            string         `json:"name"`
	Engine          entity.Engine  `json:"engine"`
	Host            string         `json:"host"`
	Port            int            `json:"port"`
	Database        string         `json:"database"`
	Username        string         `json:"username"`
	SSLMode         entity.SSLMode `json:"sslMode"`
	SSLCAPath       string         `json:"sslCaPath"`
	SSLCertPath     string         `json:"sslCertPath"`
	SSLKeyPath      string         `json:"sslKeyPath"`
	ProxyURL        string         `json:"proxyUrl"`
	SSHTunnel       SSHTunnelView  `json:"sshTunnel"`
	ReadOnly        bool           `json:"readOnly"`
	AutoReconnect   bool           `json:"autoReconnect"`
	MaxOpenConns    int            `json:"maxOpenConns"`
	MaxIdleConns    int            `json:"maxIdleConns"`
	ConnMaxLifetime int            `json:"connMaxLifetimeSeconds"`
	ConnMaxIdleTime int            `json:"connMaxIdleTimeSeconds"`
	Favorite        bool           `json:"favorite"`
	HasPassword     bool           `json:"hasPassword"`
	HasProxyAuth    bool           `json:"hasProxyCredentials"`
	CreatedAt       string         `json:"createdAt"`
	UpdatedAt       string         `json:"updatedAt"`
}

type ConnectionPatchInput struct {
	WorkspaceID     *string              `json:"workspaceId"`
	Name            *string              `json:"name"`
	Engine          *entity.Engine       `json:"engine"`
	Host            *string              `json:"host"`
	Port            *int                 `json:"port"`
	Database        *string              `json:"database"`
	Username        *string              `json:"username"`
	Password        *string              `json:"password"`
	ClearPassword   *bool                `json:"clearPassword"`
	SSLMode         *entity.SSLMode      `json:"sslMode"`
	SSLCAPath       *string              `json:"sslCaPath"`
	SSLCertPath     *string              `json:"sslCertPath"`
	SSLKeyPath      *string              `json:"sslKeyPath"`
	ProxyURL        *string              `json:"proxyUrl"`
	ClearProxyAuth  *bool                `json:"clearProxyCredentials"`
	SSHTunnel       *SSHTunnelPatchInput `json:"sshTunnel"`
	ReadOnly        *bool                `json:"readOnly"`
	AutoReconnect   *bool                `json:"autoReconnect"`
	MaxOpenConns    *int                 `json:"maxOpenConns"`
	MaxIdleConns    *int                 `json:"maxIdleConns"`
	ConnMaxLifetime *int                 `json:"connMaxLifetimeSeconds"`
	ConnMaxIdleTime *int                 `json:"connMaxIdleTimeSeconds"`
}

type SSHTunnelPatchInput struct {
	Enabled        *bool   `json:"enabled"`
	Host           *string `json:"host"`
	Port           *int    `json:"port"`
	Username       *string `json:"username"`
	Password       *string `json:"password"`
	ClearPassword  *bool   `json:"clearPassword"`
	PrivateKeyPath *string `json:"privateKeyPath"`
	KnownHostsPath *string `json:"knownHostsPath"`
}

type ConnectionTestResult struct {
	OK        bool   `json:"ok"`
	LatencyMS int64  `json:"latencyMs"`
	Message   string `json:"message"`
}

type ConnectionFavoriteInput struct {
	Favorite *bool `json:"favorite"`
}

type SSHTunnelView struct {
	Enabled        bool   `json:"enabled"`
	Host           string `json:"host"`
	Port           int    `json:"port"`
	Username       string `json:"username"`
	PrivateKeyPath string `json:"privateKeyPath"`
	KnownHostsPath string `json:"knownHostsPath"`
	HasPassword    bool   `json:"hasPassword"`
}

File: website/src/data/mappers/connection.ts
import type { ApiConnection, ApiConnectionStatus, ApiConnectionTestResult } from '../contracts/connection'
import type { Connection, ConnectionInput, ConnectionRuntimeStatus, ConnectionTestResult } from '../../entities/connection'

export function mapConnection(value: ApiConnection): Connection {
  return {
    id: value.id,
    workspaceId: value.workspaceId,
    name: value.name,
    engine: value.engine,
    host: value.host,
    port: value.port,
    database: value.database,
    username: value.username,
    sslMode: value.sslMode,
    sslCaPath: value.sslCaPath,
    sslCertPath: value.sslCertPath,
    sslKeyPath: value.sslKeyPath,
    proxyUrl: value.proxyUrl,
    sshTunnel: {
      enabled: value.sshTunnel.enabled,
      host: value.sshTunnel.host,
      port: value.sshTunnel.port,
      username: value.sshTunnel.username,
      privateKeyPath: value.sshTunnel.privateKeyPath,
      knownHostsPath: value.sshTunnel.knownHostsPath,
      hasPassword: value.sshTunnel.hasPassword,
    },
    readOnly: value.readOnly,
    autoReconnect: value.autoReconnect,
    maxOpenConns: value.maxOpenConns,
    maxIdleConns: value.maxIdleConns,
    connMaxLifetimeSeconds: value.connMaxLifetimeSeconds,
    connMaxIdleTimeSeconds: value.connMaxIdleTimeSeconds,
    favorite: value.favorite,
    hasPassword: value.hasPassword,
    hasProxyCredentials: value.hasProxyCredentials,
    status: value.status ?? 'disconnected',
    latencyMs: value.latencyMs,
    lastConnectedAt: value.lastConnectedAt,
    lastErrorCode: value.lastErrorCode,
    activeTransactions: value.activeTransactions,
    createdAt: value.createdAt,
    updatedAt: value.updatedAt,
  }
}

export function toConnectionRequest(value: ConnectionInput) {
  return {
    workspaceId: value.workspaceId,
    name: value.name,
    engine: value.engine,
    host: value.host,
    port: value.port,
    database: value.database,
    username: value.username,
    password: value.password,
    clearPassword: value.clearPassword ?? false,
    sslMode: value.sslMode,
    sslCaPath: value.sslCaPath,
    sslCertPath: value.sslCertPath,
    sslKeyPath: value.sslKeyPath,
    proxyUrl: value.proxyUrl,
    clearProxyCredentials: value.clearProxyCredentials ?? false,
    sshTunnel: {
      enabled: value.sshTunnel.enabled,
      host: value.sshTunnel.host,
      port: value.sshTunnel.port,
      username: value.sshTunnel.username,
      password: value.sshTunnel.password,
      clearPassword: value.sshTunnel.clearPassword ?? false,
      privateKeyPath: value.sshTunnel.privateKeyPath,
      knownHostsPath: value.sshTunnel.knownHostsPath,
    },
    readOnly: value.readOnly,
    autoReconnect: value.autoReconnect,
    maxOpenConns: value.maxOpenConns,
    maxIdleConns: value.maxIdleConns,
    connMaxLifetimeSeconds: value.connMaxLifetimeSeconds,
    connMaxIdleTimeSeconds: value.connMaxIdleTimeSeconds ?? 0,
  }
}

export function mapConnectionStatus(value: ApiConnectionStatus): ConnectionRuntimeStatus {
  return {
    status: value.status ?? 'disconnected',
    latencyMs: value.latencyMs,
    lastConnectedAt: value.lastConnectedAt,
    lastErrorCode: value.lastErrorCode,
    activeTransactions: value.activeTransactions,
  }
}

export function mapConnectionTest(value: ApiConnectionTestResult): ConnectionTestResult {
  return {
    ok: value.ok,
    latencyMs: value.latencyMs,
    message: value.message,
  }
}
